### PR TITLE
Opgops 2154

### PR DESCRIPTION
**Pull request by brettminnie**
Basic cloudfront support

### CODE REVIEW

**Review by tim-mcarthur-moj — APPROVED** [review]

File: roles/cloudfront/tasks/create.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+---
+
+- name: Create distribution with custom certificate
+  cloudfront:
+    type: distribution
+    state: present
+    wait_for_deployed: no
```
Comment (Azulinho) [inline roles/cloudfront/tasks/create.yml]: don't you want to wait_for_deployed: yes here ?
This considering you are doing a lookup for that cloudfront distribution just below
```  until: "{{ cf_facts['ansible_facts']['cloudfront'][item.alias_domains.0]['Distribution']['Status'] == 'Deployed' }}"```

File: roles/cloudfront/tasks/create.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    distribution: true
+    domain_name_alias: "{{ item.alias_domains.0 }}"
+  register: cf_facts
+  until: "{{ cf_facts['ansible_facts']['cloudfront'][item.alias_domains.0]['Distribution']['Status'] == 'Deployed' }}"
+  retries: 30
+  delay: 60
+
+- name: Declare cloudfront zoneid
+  set_fact:
+    cf_zone_id: Z2FDTNDATAQYW2
```
Comment (Azulinho) [inline roles/cloudfront/tasks/create.yml]: can we get that ```Z2FDTNDATAQYW2``` into a variable somewhere please?

File: roles/cloudfront/tasks/destroy.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    domain_name_alias: "{{ item.alias_domains.0 }}"
+  register: cf_facts
+
+- name: Set fact if we have have a cloudfront distribution
+  set_fact:
+    cf_exists: "{{ cf_facts['ansible_facts']['cloudfront'][item.alias_domains.0]['Distribution']['Id'] is defined }}"
+
+- block:
+    # Possibly fire and forget this, there are no sg's attached and once we remove the A
+    # records it is inacessible
```
Comment (Azulinho) [inline roles/cloudfront/tasks/destroy.yml]: I would leave that wait_for_deployed set as it is now, it means that we know where we stand by the time this play finishes

File: roles/cloudfront/tasks/create.yml (hunk 1 above)
Comment (brettminnie) [reply roles/cloudfront/tasks/create.yml]: I had it set like that originally. Provisioning time is quite varied, so did it the same way we do rds instances, poll every minute until the condition fails

File: roles/cloudfront/tasks/create.yml (hunk 2 above)
Comment (brettminnie) [reply roles/cloudfront/tasks/create.yml]: Not sure why, it's defined by aws. See here http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-route53-aliastarget.html

File: roles/cloudfront/templates/distribution.json [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        },
+        "CustomOriginConfig": {
+          "HTTPPort": 80,
+          "HTTPSPort": 443,
+          "OriginProtocolPolicy": "match-viewer",
+          "OriginSslProtocols": {
+            "Quantity": 4,
+            "Items": [
+              "SSLv3",
+              "TLSv1",
```
Comment (Azulinho) [inline roles/cloudfront/templates/distribution.json]: can we drop sslv3 and tls 1.0 ?

File: roles/cloudfront/tasks/destroy.yml (hunk 3 above)
Comment (brettminnie) [reply roles/cloudfront/tasks/destroy.yml]: I'll remove the comment, is redundant

File: roles/cloudfront/templates/distribution.json (hunk 4 above)
Comment (brettminnie) [reply roles/cloudfront/templates/distribution.json]: We're not sure what the clients are going to be using seeing as this is open to the great unwashed eventually

`dismissive hand gesture`

File: roles/cloudfront/tasks/create.yml (hunk 2 above)
Comment (Azulinho) [reply roles/cloudfront/tasks/create.yml]: just add a comment stating why and that URL then please

File: roles/cloudfront/tasks/create.yml (hunk 1 above)
Comment (Azulinho) [reply roles/cloudfront/tasks/create.yml]: won't you get a race condition where ['distribution'] may not be yet set due to eventual consistency here?

File: roles/cloudfront/tasks/destroy.yml (hunk 3 above)
Comment (Azulinho) [reply roles/cloudfront/tasks/destroy.yml]: +1

File: roles/cloudfront/templates/distribution.json (hunk 4 above)
Comment (Azulinho) [reply roles/cloudfront/templates/distribution.json]: SSLv3 and TLSv1 are broken, we should not support it.
They're as safe as using plain text

File: roles/cloudfront/tasks/create.yml (hunk 1 above)
Comment (brettminnie) [reply roles/cloudfront/tasks/create.yml]: Ok lets move it to the wait with an extended timeout, but we'll have to monitor it to see if we have guessed it big enough

File: roles/cloudfront/tasks/create.yml (hunk 2 above)
Comment (brettminnie) [reply roles/cloudfront/tasks/create.yml]: :+1:

**Pull request merged**